File: ui/flutter/lib/src/bindings/api/path/path.dart
import 'dart:io';

import 'package:lonanote/src/common/config/app_config.dart';
import 'package:lonanote/src/common/log.dart';
import 'package:path_provider/path_provider.dart';

import '../../bindings.dart';

class RustPath {
  static Future<void> initPath() async {
    final tempDir = await getTemporaryDirectory();
    final downloadDir = await getDownloadsDirectory();
    var documentDir = await getApplicationDocumentsDirectory();
    final dataDir = await getApplicationSupportDirectory();
    if (Platform.isAndroid) {
      final directory = await getExternalStorageDirectory();
      if (directory != null) {
        documentDir = directory;
      }
    }
    Bindings.invoke(key: "path.init_dir", value: {
      "dataDir": dataDir.path,
      "cacheDir": tempDir.path,
      "downloadDir": downloadDir?.path ?? documentDir.path,
      "homeDir": documentDir.path,
      "rootDir": documentDir.path,
    });
  }

  static String getCacheDir() {
    return Bindings.invoke(key: "path.get_cache_dir");
  }

  static String getHomeDir() {
    return Bindings.invoke(key: "path.get_home_dir");
  }

  static String getDataDir() {
    return Bindings.invoke(key: "path.get_data_dir");
  }

  static String getDownloadDir() {
    return Bindings.invoke(key: "path.get_download_dir");
  }

  static void test() {
    if (!AppConfig.isDebug) return;

    logger.i("RustPath.getCacheDir = ${getCacheDir()}");
    logger.i("RustPath.getHomeDir = ${getHomeDir()}");
    logger.i("RustPath.getDataDir = ${getDataDir()}");
    logger.i("RustPath.getDownloadDir = ${getDownloadDir()}");
  }
}

File: rust/core/src/api/path.rs
use cmdreg::command;

use crate::config::app_path;

#[command("path")]
fn get_cache_dir() -> String {
    app_path::get_cache_dir()
}

#[command("path")]
fn get_home_dir() -> String {
    app_path::get_home_dir()
}

#[command("path")]
fn get_data_dir() -> String {
    app_path::get_data_dir()
}

#[command("path")]
fn get_download_dir() -> String {
    app_path::get_download_dir()
}

#[command("path")]
fn init_dir(data_dir: String, cache_dir: String, download_dir: String, home_dir: String) {
    app_path::init_dir(data_dir, cache_dir, download_dir, home_dir);
}
